client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java

```diff
@@ -0,0 +1,224 @@
+     * non-recoverable failure.
+     */
+    public boolean isRunning() {
+        return isRunning;
+    }
+
+    /**
+     * When this response was created as a timestamp in milliseconds since epoch.
+     */
+    public long getStartTime() {
+        return startTimeMillis;
+    }
+
+    /**
+     * When this response will expired as a timestamp in milliseconds since epoch.
+     */
+    public long getExpirationTime() {
+        return expirationTimeMillis;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (id != null) {
+            builder.field("id", id);
+        }
+        builder.field("version", version);
+        builder.field("is_partial", isPartial);
+        builder.field("is_running", isRunning);
+        builder.field("start_time_in_millis", startTimeMillis);
+        builder.field("expiration_time_in_millis", expirationTimeMillis);
+
+        if (searchResponse != null) {
+            builder.field("response");
+            searchResponse.toXContent(builder, params);
+        }
+        if (error != null) {
+            builder.startObject("error");
+            error.toXContent(builder, params);
+            builder.endObject();
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    public static final ParseField ID_FIELD = new ParseField("id");
+    public static final ParseField VERSION_FIELD = new ParseField("version");
+    public static final ParseField IS_PARTIAL_FIELD = new ParseField("is_partial");
+    public static final ParseField IS_RUNNING_FIELD = new ParseField("is_running");
+    public static final ParseField START_TIME_FIELD = new ParseField("start_time_in_millis");
+    public static final ParseField EXPIRATION_FIELD = new ParseField("expiration_time_in_millis");
+    public static final ParseField RESPONSE_FIELD = new ParseField("response");
+    public static final ParseField ERROR_FIELD = new ParseField("error");
+
+    public static final ConstructingObjectParser<AsyncSearchResponse, Void> PARSER = new ConstructingObjectParser<>(
+            "submit_async_search_response", true,
+            args -> new AsyncSearchResponse(
+                    (int) args[0],
+                    (boolean) args[1],
+                    (boolean) args[2],
+                    (long) args[3],
+                    (long) args[4]));
+    static {
+        PARSER.declareInt(ConstructingObjectParser.constructorArg(), VERSION_FIELD);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), IS_PARTIAL_FIELD);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), IS_RUNNING_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), START_TIME_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), EXPIRATION_FIELD);
+        PARSER.declareString(AsyncSearchResponse::setId, ID_FIELD);
```

Comment:
Why can't this be a constructor arg ?

Reply:
True, this could be an `optionalConstructorArg()`. I think I did this via setters because I disklike too many optional ctor args that can be `null` and made the setters for them package private. I can even make them private I think, would that be okay with you? I don't remember exactly if more ctor args somehow slow down ConstructingObjectParser since they need to be buffered before object creation but I guess that would be a minor disadvantage.

Reply:
I pushed some changes and went with `optionalConstructorArg` to avoid more unnecessary roundtrips on this, in the end I don't care that much.